File: python/endpoints/docking/get_database.py
from clearpath_config.common.utils.yaml import read_yaml
import rclpy
from rclpy.node import Node
from clearpath_dock_msgs.srv import GetDockDatabase

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'GetDockDatabase'


class GetDockDatabaseNode(Node):

    def __init__(self, namespace: str):
        super().__init__('get_dock_database_client')
        self._get_dock_database_client = self.create_client(GetDockDatabase, f'/{namespace}/docking/dock_manager/get_database')

        while not self._get_dock_database_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/docking/dock_manager/get_database service not available, waiting again...')

    def get_database(self):
        req = GetDockDatabase.Request()
        self.future = self._get_dock_database_client.call_async(req)
        self.future.add_done_callback(self.get_database_srv_response_cb)

    def get_database_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Docks retrieved successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to retrieve docks: {response.message}')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    get_dock_database_client = GetDockDatabaseNode(namespace)
    get_dock_database_client.get_database()

    rclpy.spin(get_dock_database_client)


if __name__ == '__main__':
    main()

File: cpp/endpoints/docking/get_database.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_dock_msgs/srv/get_dock_database.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("get_dock_database_client");
  rclcpp::Client<clearpath_dock_msgs::srv::GetDockDatabase>::SharedPtr client =
    node->create_client<clearpath_dock_msgs::srv::GetDockDatabase>("/" + namespace_ + "/" + std::string("docking/dock_manager/get_database"));

  auto request = std::make_shared<clearpath_dock_msgs::srv::GetDockDatabase::Request>();

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("get_dock_database_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("get_dock_database_client"), "/%s/docking/dock_manager/get_database service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("get_dock_database_client"), "Docks retrieved successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("get_dock_database_client"), "Failed to retrieve docks: %s", result.get()->message.c_str());
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("get_dock_database_client"), "Failed to call service /%s/docking/dock_manager/get_database", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}
